File: ruby/ruby/Ice/Proxy.rb
# Copyright (c) ZeroC, Inc.

module Ice

    T_ObjectPrx = Ice.__declareProxy('::Ice::Object')
    T_ObjectPrx.defineProxy(ObjectPrx, nil, [])

    # Provide some common functionality for proxy classes
    module Proxy_mixin
        module ClassMethods
            def ice_staticId()
                self::ICE_ID
            end

            def checkedCast(proxy, facet: nil, context: nil)
                if proxy.nil?
                    return nil
                end

                unless facet.nil?
                    proxy = proxy.ice_facet(facet)
                end

                if proxy.ice_isA(self::ICE_ID, context)
                    return self::create(proxy)
                else
                    return nil
                end
            end

            def uncheckedCast(proxy, facet: nil)
                if proxy.nil?
                    return nil
                end

                unless facet.nil?
                    proxy = proxy.ice_facet(facet)
                end

                self::create(proxy)
            end
        end

      def self.included(base)
          base.extend(ClassMethods)
      end
    end

    class ObjectPrx
        include Proxy_mixin
    end

    # Proxy comparison functions.

    def Ice.proxyIdentityCompare(lhs, rhs)
      if (lhs && !lhs.is_a?(ObjectPrx)) || (rhs && !rhs.is_a?(ObjectPrx))
          raise TypeError, 'argument is not a proxy'
      end
      if lhs.nil? && rhs.nil?
          return 0
      elsif lhs.nil? && rhs
          return -1
      elsif lhs && rhs.nil?
          return 1
      else
          return lhs.ice_getIdentity() <=> rhs.ice_getIdentity()
      end
    end

    def Ice.proxyIdentityEqual(lhs, rhs)
        return proxyIdentityCompare(lhs, rhs) == 0
    end

    def Ice.proxyIdentityAndFacetCompare(lhs, rhs)
        n = proxyIdentityCompare(lhs, rhs)
        if n == 0 && lhs && rhs
            n = lhs.ice_getFacet() <=> rhs.ice_getFacet()
        end
        return n
    end

    def Ice.proxyIdentityAndFacetEqual(lhs, rhs)
        return proxyIdentityAndFacetCompare(lhs, rhs) == 0
    end
end

File: php/lib/Ice/Proxy.php
<?php
// Copyright (c) ZeroC, Inc.

namespace Ice;

class ObjectPrxHelper
{
    public static function createProxy($communicator, $proxyString)
    {
        return $communicator->stringToProxy($proxyString, '::Ice::Object');
    }

    public static function checkedCast($proxy, ...$args)
    {
        return $proxy->ice_checkedCast('::Ice::Object', ...$args);
    }

    public static function uncheckedCast($proxy, $facet=null)
    {
        return $proxy->ice_uncheckedCast('::Ice::Object', $facet);
    }

    public static function ice_staticId()
    {
        return '::Ice::Object';
    }
}

global $Ice__t_ObjectPrx;
$Ice__t_ObjectPrx = IcePHP_defineProxy('::Ice::Object', null, null);

//
// Proxy comparison functions.
//
function proxyIdentityCompare($lhs, $rhs)
{
    if (($lhs != null && !($lhs instanceof ObjectPrx)) || ($rhs != null && !($rhs instanceof ObjectPrx))) {
        throw new InvalidArgumentException('argument is not a proxy');
    }
    if ($lhs == null && $rhs == null) {
        return 0;
    } elseif ($lhs == null && $rhs != null) {
        return -1;
    } elseif ($lhs != null && $rhs == null) {
        return 1;
    } else {
        $lid = $lhs->ice_getIdentity();
        $rid = $rhs->ice_getIdentity();
        if ($lid < $rid) {
            return -1;
        } elseif ($lid > $rid) {
            return 1;
        } else {
            return 0;
        }
    }
}

function proxyIdentityEqual($lhs, $rhs)
{
    return proxyIdentityCompare($lhs, $rhs) == 0;
}

function proxyIdentityAndFacetCompare($lhs, $rhs)
{
    $n = proxyIdentityCompare($lhs, $rhs);
    if ($n == 0 && $lhs != null && $rhs != null) {
        $n = strcmp($lhs->ice_getFacet(), $rhs->ice_getFacet());
    }
    return $n;
}

function proxyIdentityAndFacetEqual($lhs, $rhs)
{
    return proxyIdentityAndFacetCompare($lhs, $rhs) == 0;
}

?>
